content/recipes/async-local-storage.md

```diff
@@ -0,0 +1,227 @@
+### Async Local Storage
+
+`AsyncLocalStorage` is a [Node.js API](https://nodejs.org/api/async_context.html#async_context_class_asynclocalstorage) (based on the `async_hooks` API) that provides an alternative way of propagating local state through the application without the need to explicitly pass it as a function parameter. It is similar to a thread-local storage in other languages.
+
+The main idea of Async Local Storage is that we can _wrap_ some function call with the `AsyncLocalStorage#run` call. All code that is invoked within the wrapped call gets access to the same `store`, which will be unique to each call chain.
+
+In the context of NestJS, that means if we can find a place within the request's lifecycle where we can wrap the rest of the request's code, we will be able to access and modify state visible only to that request, effectively eliminating the need for REQUEST scoped providers and their limitations.
+
+#### Custom implementation
+
+NestJS itself does not provide any built-in abstraction for `AsyncLocalStorage`, so let's walk through how we could implement it ourselves for the simplest HTTP case to get a better understanding of the whole concept:
+
+> info **info** For a ready-made [dedicated package](recipes/async-local-storage#nestjs-cls), continue reading below.
+
+1. First, create a new instance of the `AsyncLocalStorage` in some shared source file. Since we're using NestJS, let's also turn it into a module with a custom provider.
+
+```ts
+/** als.module.ts */
+
+import { AsyncLocalStorage } from 'async_hooks';
+import { Module } from '@nestjs/core';
+
+export const asyncLocalStorage = new AsyncLocalStorage();
```

Comment:
not sure if the instance will be needed somewhere

If not, then we can move this to L29 instead, turn it into an in-line version

Reply:
I thought about that also. But we can use the instance outside of the DI context if needed. But it's an advanced use case (and possibly an anti pattern), so yeah, let's not expose it in the example.

Reply:
I guess I could also put `providers: [AsyncLocalStorage]` directly and NestJS would automatically create an instace, but I wanted to be explicit here.